docker_fargate/docker_fargate_stack.py

```diff
@@ -129,6 +131,21 @@ def __init__(self, scope: Construct, context: str, env: dict, vpc: ec2.Vpc, **kw
 
             # Other metrics to drive scaling are discussed here:
             # https://docs.aws.amazon.com/cdk/api/v1/python/aws_cdk.aws_autoscaling/README.html
+            # Add more capacity when active connections increase
+            active_connection_metric = cloudwatch.Metric(
+                namespace = "AWS/ApplicationELB",
+                metric_name="ActiveConnectionCount",
+                period=Duration.seconds(60),
+                statistic="AVERAGE",
+                unit=cloudwatch.Unit("COUNT")
+            )
+
+            scalable_target.scale_on_metric("ScaleToActiveConnection",
+                metric=active_connection_metric,
+                scaling_steps=[autoscaling.ScalingInterval(lower=3, change=+1), autoscaling.ScalingInterval(lower=6, change=+2), autoscaling.ScalingInterval(lower=9, change=+3)
```

Comment:
I'm not sure this logic will do what we want. If I'm understanding this right, then this will add tasks to the cluster any time there are more than 3 active connections - and looking at the `app/dca-p-dcapr-1WUB7ZA0SUK77/69f780000a3a72f6` metric in dca-prod, I see that the metric rarely goes below 3. So this means that the dca-prod cluster would continuously grow to its max size while maintaining a stable connection count of 3, right? Also, I don't see a scale-down configuration, are we relying on the memory usage scaling to reduce the capacity when it's no longer needed?

Reply:
My impression is that [min_capacity=2](https://github.com/Sage-Bionetworks/data_curator-infra/blob/d096074e777a407bae4d488612ededab315cb356/docker_fargate/docker_fargate_stack.py#L116) would keep a baseline of two instances and that 3-5 connections would add one instance instead of scaling to the `max_capacity`. Then 6-8 connections would add another. The scaling should only happen once per threshold crossing, not continuously.

I didn't consider a scale-down configuration so I will add that. I imagine that without a scale-down config, if the alarm toggles frequently then it will keep adding instances up to the max, but will never remove them.

Once getting this to work in dev, I'll adjust the properties for prod. Which brings up another point of refactoring this to be different for each deployment context. [The alarm for this metric dev](https://us-east-1.console.aws.amazon.com/cloudwatch/home?region=us-east-1#alarmsV2:alarm/dca-active-connection?) isn't always going off, so I think 3 would work. At this point I'd rather set the scaling threshold lower to test how it reacts. Then we can adjust the thresholds and min/max capacities later.

Reply:
[Reading the aws_applicationautoscaling docs](https://github.com/aws/aws-cdk/tree/33c820b2fe16e34d52e71bed7e1ef598f62f0bd2/packages/aws-cdk-lib/aws-applicationautoscaling) I think you're right. This would add a new instance every evaluation period (1 min) if the threshold is triggered.

Reply:
Oh, you might be right, I don't see anywhere in the documentation that specifies if the step scaling is being applied to the baseline or to the current capacity. I'm +1 on testing/tuning in dev to find out.